identity_iota_core/src/document/iota_document.rs

```diff
@@ -275,23 +277,52 @@ impl IotaDocument {
     StateMetadataDocument::from(self).pack(encoding)
   }
 
-  /// Deserializes the document from the state metadata bytes of an Alias Output.
+  /// Deserializes the document from an Alias Output.
   ///
   /// If `allow_empty` is true, this will return an empty DID document marked as `deactivated`
   /// if `state_metadata` is empty.
   ///
   /// NOTE: `did` is required since it is omitted from the serialized DID Document and
   /// cannot be inferred from the state metadata. It also indicates the network, which is not
   /// encoded in the `AliasId` alone.
-  pub fn unpack(did: &IotaDID, state_metadata: &[u8], allow_empty: bool) -> Result<IotaDocument> {
-    if state_metadata.is_empty() && allow_empty {
+  pub fn unpack_from_output(did: &IotaDID, alias_output: &AliasOutput, allow_empty: bool) -> Result<IotaDocument> {
+    if alias_output.state_metadata().is_empty() && allow_empty {
       let mut empty_document = IotaDocument::new_with_id(did.clone());
       empty_document.metadata.created = None;
       empty_document.metadata.updated = None;
       empty_document.metadata.deactivated = Some(true);
       return Ok(empty_document);
     }
-    StateMetadataDocument::unpack(state_metadata).and_then(|doc| doc.into_iota_document(did))
+    let mut iota_document: IotaDocument =
+      StateMetadataDocument::unpack(alias_output.state_metadata()).and_then(|doc| doc.into_iota_document(did))?;
+    iota_document.set_controller_and_governor_addresses(alias_output, &did.network_str().to_owned().try_into()?);
+    Ok(iota_document)
+  }
+
+  fn set_controller_and_governor_addresses(&mut self, alias_output: &AliasOutput, network_name: &NetworkName) {
+    self.metadata.governor_address = Some(*alias_output.governor_address());
+    self.metadata.state_controller_address = Some(*alias_output.state_controller_address());
+
+    let controller_did: Option<IotaDID> = {
+      match alias_output.unlock_conditions().state_controller_address() {
+        Some(unlock_condition) => match unlock_condition.address() {
+          Address::Alias(alias_address) => Some(IotaDID::new(alias_address.alias_id(), network_name)),
+          Address::Nft(_) | Address::Ed25519(_) => None,
+        },
+        None => None,
+      }
+    };
+    if let Some(controller_did) = controller_did {
+      let controller: &mut Option<OneOrSet<IotaDID>> = self.core_document_mut().controller_mut();
+      match controller {
+        Some(controller) => {
+          controller.append(controller_did);
+        }
+        None => {
+          *controller = Some(OneOrSet::new_one(controller_did));
+        }
+      }
+    }
```

Comment:
Overwrite the controller completely, any other controller that's not a state controller of the Alias Output isn't really a controller for the IOTA DID Method, in my opinion.

```suggestion
    // Overwrite the DID Document controller.
    *self.core_document_mut().controller_mut() = controller_did.map(|did| OneOrSet::new_one(did));
```